pkg/planner/core/stats.go

```diff
@@ -130,6 +130,15 @@ func deriveStats4DataSource(lp base.LogicalPlan) (*property.StatsInfo, bool, err
 	for i, expr := range ds.PushedDownConds {
 		ds.PushedDownConds[i] = expression.EliminateNoPrecisionLossCast(exprCtx, expr)
 	}
+	// Prune indexes based on WHERE and ordering columns (ORDER BY, MIN/MAX/FIRST_VALUE) if we have too many
+	threshold := ds.SCtx().GetSessionVars().OptIndexPruneThreshold
+	if len(ds.AllPossibleAccessPaths) > threshold && (len(ds.WhereColumns) > 0 || len(ds.OrderingColumns) > 0 || len(ds.JoinColumns) > 0) {
```

Comment:
Almost LGTM, but just a suggestion that could we implement this as a separate logical rule like `rule/large_number_index_pruning`? And then converge all logic of this process like collecting interesting columns, pruning indexes and scoring indexes into that rule, which should make this feature easier to maintain in the future.